packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -232,10 +233,139 @@ export default createRule<Options, MessageId>({
+      for (const argument of node.arguments) {
+        if (argument.type === AST_NODE_TYPES.SpreadElement) {
+          break;
+        }
+
+        checkableArguments.push(argument);
+      }
+
+      // nothing to do
+      if (checkableArguments.length === 0) {
+        return undefined;
+      }
+
+      // Game plan: we're going to check the type of the callee. If it has call
+      // signatures and they _ALL_ agree that they assert on a parameter at the
+      // _SAME_ position, we'll consider the argument in that position to be an
+      // asserted argument.
+      const calleeType = getConstrainedTypeAtLocation(services, node.callee);
+      const callSignatures = tsutils.getCallSignaturesOfType(calleeType);
+
+      let assertedParameterIndex: number | undefined = undefined;
+      for (const signature of callSignatures) {
+        const declaration = signature.getDeclaration();
+        const returnTypeAnnotation = declaration.type;
+
+        // Be sure we're dealing with a truthiness assertion function.
+        if (
+          !(
+            returnTypeAnnotation != null &&
+            ts.isTypePredicateNode(returnTypeAnnotation) &&
+            // This eliminates things like `x is string` and `asserts x is T`
+            // leaving us with just the `asserts x` cases.
+            returnTypeAnnotation.type == null &&
+            // I think this is redundant but, still, it needs to be true
+            returnTypeAnnotation.assertsModifier != null
+          )
+        ) {
+          return undefined;
+        }
+
+        const assertionTarget = returnTypeAnnotation.parameterName;
+        if (assertionTarget.kind !== ts.SyntaxKind.Identifier) {
+          // This can happen when asserting on `this`. Ignore!
+          return undefined;
+        }
+
+        // If the first parameter is `this`, skip it, so that our index matches
+        // the index of the argument at the call site.
+        const firstParameter = declaration.parameters.at(0);
+        const nonThisParameters =
+          firstParameter?.name.kind === ts.SyntaxKind.Identifier &&
+          firstParameter.name.text === 'this'
+            ? declaration.parameters.slice(1)
+            : declaration.parameters;
+
+        // Don't bother inspecting parameters past the number of
+        // arguments we have at the call site.
+        const checkableNonThisParameters = nonThisParameters.slice(
+          0,
+          checkableArguments.length,
+        );
+
+        let assertedParameterIndexForThisSignature: number | undefined;
+        for (const [index, parameter] of checkableNonThisParameters.entries()) {
+          if (parameter.dotDotDotToken != null) {
+            // Cannot assert a rest parameter, and can't have a rest parameter
+            // before the asserted parameter. It's not only a TS error, it's
+            // not something we can logically make sense of, so give up here.
+            return undefined;
+          }
```

Comment:
[probably dead code] The rest parameter can only be the last parameter, so the asserted parameter *should* be before the rest parameter (otherwise `asserts` must assert the name of the rest parameter, but that's a TS error). I can't think of a single where this condition would be true in the semantically valid TS code.

Reply:
> I can't think of a single where this condition would be true in the semantically valid TS code.

Agreed, that's what I tried to express in the comment. But I still think it's worth ensuring the rule doesn't crash. Added some test cases that hit this.